File: source_go/internal/cloud/dropbox.go
package cloud

import (
    "bytes"
    "encoding/json"
    "fmt"
    "net/http"
    "path/filepath"
    "time"
)

type DropboxProvider struct {
    AccessToken string
    Client      *http.Client
}

func NewDropboxProvider(token string) *DropboxProvider {
    return &DropboxProvider{
        AccessToken: token,
        Client:      &http.Client{},
    }
}

func (p *DropboxProvider) Name() string {
    return "dropbox"
}

func (p *DropboxProvider) request(endpoint string, body interface{}) (map[string]interface{}, error) {
    jsonBody, err := json.Marshal(body)
    if err != nil {
        return nil, err
    }

    req, err := http.NewRequest("POST", "https://api.dropboxapi.com/2/files/"+endpoint, bytes.NewBuffer(jsonBody))
    if err != nil {
        return nil, err
    }

    req.Header.Set("Authorization", "Bearer "+p.AccessToken)
    req.Header.Set("Content-Type", "application/json")

    resp, err := p.Client.Do(req)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    if resp.StatusCode != http.StatusOK {
        return nil, fmt.Errorf("dropbox api error: %s", resp.Status)
    }

    var result map[string]interface{}
    if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
        return nil, err
    }
    return result, nil
}

func (p *DropboxProvider) ListFiles(path string) ([]CloudFile, error) {
    var files []CloudFile
    hasMore := true
    var cursor string

    if path == "." || path == "/" {
        path = ""
    }

    for hasMore {
        var body map[string]interface{}
        endpoint := "list_folder"

        if cursor != "" {
            endpoint = "list_folder/continue"
            body = map[string]interface{}{"cursor": cursor}
        } else {
            body = map[string]interface{}{
                "path":                                path,
                "recursive":                           true,
                "include_media_info":                  false,
                "include_deleted":                     false,
                "include_has_explicit_shared_members": false,
            }
        }

        res, err := p.request(endpoint, body)
        if err != nil {
            return nil, err
        }

        if entries, ok := res["entries"].([]interface{}); ok {
            for _, e := range entries {
                entry := e.(map[string]interface{})
                if tag, ok := entry[".tag"].(string); ok && tag == "file" {
                    name := entry["name"].(string)
                    pathDisplay := entry["path_display"].(string)
                    id := entry["id"].(string)
                    size := int64(entry["size"].(float64))
                    hash := ""
                    if h, ok := entry["content_hash"].(string); ok {
                        hash = h
                    }

                    // Client modified
                    modTime := time.Now()
                    if cm, ok := entry["client_modified"].(string); ok {
                        if t, err := time.Parse(time.RFC3339, cm+"Z"); err == nil { // Dropbox sends UTC without Z sometimes? Docs say ISO8601
                            modTime = t
                        }
                    }

                    files = append(files, CloudFile{
                        ID:           id,
                        Name:         name,
                        Path:         pathDisplay,
                        Hash:         hash,
                        Size:         size,
                        ModifiedTime: modTime,
                        Provider:     "dropbox",
                    })
                }
            }
        }

        if hm, ok := res["has_more"].(bool); ok {
            hasMore = hm
        } else {
            hasMore = false
        }

        if c, ok := res["cursor"].(string); ok {
            cursor = c
        }
    }

    return files, nil
}

func (p *DropboxProvider) RenameFile(file *CloudFile, newName string) error {
    parent := filepath.Dir(file.Path)
    newPath := filepath.Join(parent, newName)
    // Ensure slash
    if newPath[0] != '/' {
        newPath = "/" + newPath
    }

    body := map[string]interface{}{
        "from_path":  file.Path,
        "to_path":    newPath,
        "autorename": false,
    }

    _, err := p.request("move_v2", body)
    return err
}

func (p *DropboxProvider) DeleteFile(file *CloudFile) error {
    body := map[string]interface{}{
        "path": file.Path,
    }
    _, err := p.request("delete_v2", body)
    return err
}

File: src/cloud/dropbox.rs
use anyhow::{Result, anyhow};
use super::{CloudFile, CloudProvider};
use std::time::{SystemTime, UNIX_EPOCH};
use serde_json::Value;

pub struct DropboxProvider {
    access_token: String,
    client: reqwest::blocking::Client,
}

impl DropboxProvider {
    pub fn new(access_token: String) -> Self {
        Self {
            access_token,
            client: reqwest::blocking::Client::new(),
        }
    }

    fn request(&self, endpoint: &str, body: &Value) -> Result<Value> {
        let res = self.client.post(format!("https://api.dropboxapi.com/2/files/{}", endpoint))
            .header("Authorization", format!("Bearer {}", self.access_token))
            .header("Content-Type", "application/json")
            .json(body)
            .send()?;

        if !res.status().is_success() {
            return Err(anyhow!("Dropbox API error: {}", res.text()?));
        }

        let json: Value = res.json()?;
        Ok(json)
    }
}

impl CloudProvider for DropboxProvider {
    fn name(&self) -> &str {
        "dropbox"
    }

    fn list_files(&self, path: &str) -> Result<Vec<CloudFile>> {
        let mut files = Vec::new();
        let mut has_more = true;
        let mut cursor = None;

        let path = if path == "." || path == "/" { "" } else { path };

        while has_more {
            let body = if let Some(ref c) = cursor {
                 serde_json::json!({ "cursor": c })
            } else {
                 serde_json::json!({
                    "path": path,
                    "recursive": true,
                    "include_media_info": false,
                    "include_deleted": false,
                    "include_has_explicit_shared_members": false
                })
            };

            let endpoint = if cursor.is_some() { "list_folder/continue" } else { "list_folder" };
            let json = self.request(endpoint, &body)?;

            if let Some(entries) = json["entries"].as_array() {
                for entry in entries {
                    if entry[".tag"] == "file" {
                        let name = entry["name"].as_str().unwrap_or_default().to_string();
                        let path_display = entry["path_display"].as_str().unwrap_or_default().to_string();
                        let id = entry["id"].as_str().unwrap_or_default().to_string();
                        let size = entry["size"].as_u64().unwrap_or(0);
                        let hash = entry["content_hash"].as_str().map(|s| s.to_string());

                        // Parse client_modified
                        let modified_str = entry["client_modified"].as_str().unwrap_or("");
                        let modified_time = chrono::DateTime::parse_from_rfc3339(&format!("{}Z", modified_str))
                            .map(|dt| SystemTime::from(dt))
                            .unwrap_or(SystemTime::now());

                        files.push(CloudFile {
                            id,
                            name,
                            path: path_display,
                            hash,
                            size,
                            modified_time,
                            provider: "dropbox".to_string(),
                        });
                    }
                }
            }

            has_more = json["has_more"].as_bool().unwrap_or(false);
            cursor = json["cursor"].as_str().map(|s| s.to_string());
        }

        Ok(files)
    }

    fn rename_file(&self, file: &CloudFile, new_name: &str) -> Result<()> {
        // Calculate new path
        let parent = std::path::Path::new(&file.path).parent().unwrap_or(std::path::Path::new(""));
        let new_path = parent.join(new_name).to_str().unwrap().to_string();

        // Dropbox paths must start with /
        let new_path = if !new_path.starts_with('/') {
            format!("/{}", new_path)
        } else {
            new_path
        };

        let body = serde_json::json!({
            "from_path": file.path,
            "to_path": new_path,
            "autorename": false
        });

        self.request("move_v2", &body)?;
        Ok(())
    }

    fn delete_file(&self, file: &CloudFile) -> Result<()> {
        let body = serde_json::json!({
            "path": file.path
        });
        self.request("delete_v2", &body)?;
        Ok(())
    }
}
